### PR TITLE
error: print non-printable chars in hex style for duplicate entry error

### LINKED ISSUE #35289: Error message of duplicate primary key is not shown properly
## Enhancement

Actual:
```sql
mysql> CREATE TABLE t1 (a BINARY(16) PRIMARY KEY);
Query OK, 0 rows affected (0.02 sec)

mysql> INSERT INTO t1 VALUES (AES_ENCRYPT('a','a'));
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t1 VALUES (AES_ENCRYPT('a','a'));
ERROR 1062 (23000): Duplicate entry '{ W]�u����.�٧t' for key 'PRIMARY'
```

Expected:
```sql
mysql> CREATE TABLE t1 (a BINARY(16) PRIMARY KEY);
Query OK, 0 rows affected (0.05 sec)

mysql> INSERT INTO t1 VALUES (AES_ENCRYPT('a','a'));
Query OK, 1 row affected (0.02 sec)

mysql> INSERT INTO t1 VALUES (AES_ENCRYPT('a','a'));
ERROR 1062 (23000): Duplicate entry '{ W]\xA1\x06u\x9D\xBD\xB1\xA3.\xE2\xD9\xA7t' for key 't1.PRIMARY'
```

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve?

Issue Number: close #35289 

- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: store/driver/txn/error.go [hunk 1]
```diff
@@ -100,6 +125,8 @@ func extractKeyExistsErrFromHandle(key kv.Key, value []byte, tblInfo *model.Tabl
 		if col.Length > 0 && len(str) > col.Length {
 			str = str[:col.Length]
 		}
+
+		str = fmtNonASCIIPrintableCharToHex(str)
```
Comment (tangenta) [inline store/driver/txn/error.go:129]: We need to convert it only if the column type is binary. See `types.IsBinaryStr()`.

```sql
mysql> insert into t1 values ('测试');
Query OK, 1 row affected (0.06 sec)

mysql> insert into t1 values ('测试');
ERROR 1062 (23000): Duplicate entry '测试' for key 't1.PRIMARY'
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.27    |
+-----------+
1 row in set (0.02 sec)
```

**Comment by winkyao** [comment]
@tangenta Addressed, PTAL

**Review by tangenta — APPROVED** [review]

File: expression/integration_test.go [hunk 2]
```diff
@@ -7895,3 +7895,17 @@ func TestIssue39146(t *testing.T) {
 	tk.MustExec("set @@tidb_enable_vectorized_expression = off;")
 	tk.MustQuery(`select str_to_date(substr(dest,1,6),'%H%i%s') from sun;`).Check(testkit.Rows("20:23:10"))
 }
+
+func TestIssue35289(t *testing.T) {
```
Comment (hawkingrei) [inline expression/integration_test.go:7899]: ```TestIssue35289``` is a bad name of test cases. we should avoid it. 

I think it is a corner case about insert and we can put it into ```TestInsert``` and comment the issue id.

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**